funfact/lang/interpreter/_penalty.py

```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env python
```

Comment:
Since the penalty evaluator only operates on leaf nodes, we can simply use `dfs_filter` without defining a full-fledged interpreter. This could then be implemented as a standalone function.

Reply:
Makes sense, although the interpeter isn't doing much more work compared to a `dfs_filter`. This was modeled after the leaf initializer. Probably the same comment applies to that interpreter as well?

Reply:
I see your point here. There is a key difference between the leaf initializer and the penalty evaluator: the former modifies (insert the `data` attribute to) the AST nodes, while the latter only performs read-only operations. That's why the leaf initializer could benefit from the TranscribeInterpreter infrastructure, while the penalty evaluator can be further simplified. :sunglasses:

Reply:
Makes sense, it is replaced now by a `dfs_filter` in the factorization model. In my preliminary tests this seems to improve the performance as well 😄